src/main/java/org/openhab/tools/analysis/checkstyle/ThingHandlerSetLabelCheck.java

```diff
@@ -0,0 +1,134 @@
+            if (isHandlerExtendingBaseThingHandler(ast) && isPackageCorrect) {
+                checkSetLabelMethodCall(ast);
+            }
+        }
+    }
+
+    // Pre-order traversal - starting from the root first
+    private void checkSetLabelMethodCall(DetailAST ast) {
+        if (ast == null) {
+            return;
+        }
+
+        // Check the data part of the current node
+        if (THING.equals(ast.getText())) {
+            // thing.setLabel(...)
+            boolean isThingFieldSetLabelMethodCalled = ast.getNextSibling() != null
+                    && ast.getNextSibling().getText().equals(SET_LABEL);
+            // when literal THIS is used the setLabel method is sibling of the parent of the node - thing
+            // this.thing.setLabel(...)
+            boolean isReferredThingFieldSetLabelMethodCalled = ast.getParent().getNextSibling() != null
+                    && ast.getParent().getNextSibling().getText().equals(SET_LABEL);
+            boolean isSetLabelCalled = isThingFieldSetLabelMethodCalled || isReferredThingFieldSetLabelMethodCalled;
+
+            if (isSetLabelCalled) {
+                log(ast.getLineNo(), WARNING_MESSAGE);
+            }
+        }
+
+        // Traverse the left subtree
+        checkSetLabelMethodCall(ast.getFirstChild());
+        // Traverse the right subtree
+        checkSetLabelMethodCall(ast.getNextSibling());
+    }
+
+    // In-order traversal - to match the fully qualified name in exact order
+    // In this case - need to find node "org" first
+    // org.eclipse... etc.
+    private String getFullyQualifiedNameFromAst(DetailAST ast) {
+        if (ast == null) {
+            return "";
+        }
+
+        StringBuilder strBuilder = new StringBuilder();
+        // Traverse the left subtree
+        String firstChildString = getFullyQualifiedNameFromAst(ast.getFirstChild());
+        strBuilder.append(firstChildString);
+
+        // Check the data part of the current node
+        if (ast.getType() == TokenTypes.IDENT) {
+            if (ast.getPreviousSibling() != null) {
+                strBuilder.append(ast.getParent().getText()).append(ast.getText());
+            } else {
+                strBuilder.append(ast.getText());
+            }
+        }
+
+        // Traverse the right subtree
+        String nextSiblingString = getFullyQualifiedNameFromAst(ast.getNextSibling());
+        strBuilder.append(nextSiblingString);
+        return strBuilder.toString();
+    }
+
+    private boolean isHandlerExtendingBaseThingHandler(DetailAST ast) {
+        DetailAST extendsClause = ast.findFirstToken(TokenTypes.EXTENDS_CLAUSE);
+
+        if (extendsClause != null) {
+            String inheritedClassName = extendsClause.getFirstChild().getText();
+            // Only the simple name of the class can be retrieved from the extends clause
+            if (inheritedClassName.equals(BASE_THING_HANDLER)) {
```

Comment:
Does this work if you extend from the class indirectly? Also the import check looks non-robust to that. WDYT?

Reply:
That's a great idea. I will check this case.